Fix `eth_simulateV1` after Jovian fork

kustrun opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

After the Jovian fork, all `eth_simulateV1` calls fail with the error: `"error calculating DA footprint: missing deposit transaction"`.

The root cause is in `CalcDAFootprint()`, which assumes that a deposit transaction exists at `txs[0]` and that it contains the `daFootprintGasScalar` data. This value is required to calculate the rollup gas costs.

The proposed solution modifies **ONLY** `eth_simulateV1` by injecting a synthetic Jovian deposit transaction before the user-provided simulation transactions. This allows `CalcDAFootprint()` to succeed while keeping the core block-processing logic unchanged, ensuring that simulations work without affecting normal execution.

kustrun committed:
fix(eth_simulatev1): Create synthetic Jovian Deposit TX, to not fail in the CalcDAFootprint() function.

kustrun committed:
fix(eth_simulatev1): Return block without synthetic Jovian deposit tx.

geoknee commented on internal/ethapi/simulate.go:
```diff
@@ -359,12 +361,37 @@ func (sim *simulator) processBlock(ctx context.Context, block *simBlock, header,
+	if isJovian {
+		depositTx := createSimulatedJovianDepositTx()
+		finalTxes = append([]*types.Transaction{depositTx}, txes...)
+	}
```
If you look at this test code: 
https://github.com/ethereum-optimism/op-geth/blob/472228745b7e5c6d777897c83067fe9dacc47ad3/miner/miner_optimism_test.go#L116-L118

you can see we don't insert a jovian deposit tx until the block after the activation block.

Also, you might be able to move the helper function from the test so that it can be used here as well?

kustrun replied:
Thank you for pointing this out. I have applied the fix to resolve this.

kustrun committed:
fix(eth_simulatev1): Extract JovianDepositTx() from tests, activate Jovian fork based on parent.Time.

geoknee commented on core/types/deposit_tx.go:
```diff
@@ -117,3 +118,10 @@ func (tx *Transaction) From() common.Address {
 	return tx.inner.(interface{ from() common.Address }).from()
 }
+
+func JovianDepositTx(daFootprintGasScalar uint16) *DepositTx {
```
Although I did suggest you promote this function out of a test, I am a bit worried about having it here, especially with no godoc. This function is preparing a stubbed out / test L1 attributes transaction which may make sense in testing (and _maybe_ in simulation, although I am beginning to doubt that) but definitely not in general in production code. See other comment for some other alternative ideas @joshklop suggested.

geoknee commented on internal/ethapi/simulate.go:
```diff
@@ -359,12 +360,38 @@ func (sim *simulator) processBlock(ctx context.Context, block *simBlock, header,
+
+	// For Jovian blocks, reconstruct the block without the synthetic deposit transaction
+	// to maintain consistent indexing between transactions and receipts.
+	// We must use types.NewBlock to recompute TxHash correctly for the user transactions only.
```
I discussed this with @joshklop and we agreed there may be some better alternatives here.

In general to run a realistic simulation against an opstack chain, there should be an l1 attributes (deposit) transaction as the 0th transaction in any given block. That is currently missing from the simulation code and should be fixed:

Options:

A) This PR so far. Inject a stubbed out transaction with some arbitrary constants, only for jovian blocks, set so that newly introduced code anticipating real optimism blocks doesn't error out when it hits a block without such a transaction.

B) Require the user to provide the exact l1 attributes transaction in the `eth_simulate` call. This would break a lot of tooling most likely.

C) Have the code read the most recent l1 attributes transaction from the DB and use that. This would be accurate (apart from the `sequenceNumber`) in most cases, apart from when there the l1 origin would be advanced over an L1 block with a `SystemConfig` event, or when we cross a hardfork activation.

D) Work around the `CalcDAFootprint` error (using some escape hatch or other). This option gives up on realism in the name of pragmatism.

geoknee replied:
So far C seems like the best option in the list IMO.

geoknee replied:
I went ahead and implemented C. A nice side effect is that now `eth_simulateV1` can be used to compute the DA footprint of a transaction (inspecting the blob gas used field of the returned block simulation result).

geoknee commented on internal/ethapi/simulate.go:
```diff
@@ -359,12 +360,38 @@ func (sim *simulator) processBlock(ctx context.Context, block *simBlock, header,
+	finalTxes := txes
+	if isJovian {
+		// Use a reasonable default DA footprint gas scalar (e.g., 1 wei)
+		depositTx := types.NewTx(types.JovianDepositTx(1))
```
Using `1` for the DA Footprint Gas Scalar seems a bit arbitrary. Other alternatives:
* `0`
*  `400` (current default, but it could get stale). 
* value from previous block (see other comment)

In general the simulation result can be affected by this scalar (if it is high, it might push the transactions over the block gas limit).

geoknee commented:
Could you comment about your use case here? In what context are you running these simulations?

RealJohnnyTime commented:
+1
This is critical for accurately simulating transactions in backrunning strategies on OP Stack–based EVM chains. Many existing backrunners are currently broken because of this.

kustrun commented:
@geoknee apologies for the delayed reply, was AFK. As @RealJohnnyTime mentioned, any transaction simulation currently fails (my specific use case is tx bundle simulation). Below is an example payload that reproduces the issue, in case it helps with diagnosing and resolving it.

For now, I'm using a quick workaround with an artificial Jovian transaction until we settle on a proper solution.

```bash
  curl -X POST \
    -H "Content-Type: application/json" \
    --data '{
      "jsonrpc": "2.0",
      "id": 1,
      "method": "eth_simulateV1",
      "params": [
        {
          "blockStateCalls": [
            {
              "stateOverrides": {
                "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045": {
                  "balance": "0x21e19e0c9bab2400000"
                }
              },
              "calls": [
                {
                  "from": "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
                  "to": "0x4200000000000000000000000000000000000015",
                  "value": "0x0"
                }
              ]
            }
          ],
          "validation": false
        },
        "latest"
      ]
    }' \
    <YOUR_OP_STACK_RPC_URL>
  ```

geoknee commented:
Thanks for the context @kustrun. I'll take another quick look and may push some commits to your branch.

geoknee committed:
Inject L1 attributes tx in Optimism simulation

geoknee committed:
Unexport JovianDepositTx and move to tests

Remove exported JovianDepositTx from core/types and add a local
jovianDepositTx helper in miner tests. Also remove the unused
encoding/binary import from the core/types file and add it to the test
files where needed.

geoknee committed:
Merge remote-tracking branch 'origin/optimism' into fix/simulate_v1

geoknee commented:
/ci authorize [f2b3a69](https://github.com/ethereum-optimism/op-geth/pull/732/commits/f2b3a691667c7e062bf3e6c5e9ed4aa9339fd314)

geoknee committed:
Pass nil options to sim.execute in test

geoknee committed:
Pass nil as second argument in simulation test

geoknee commented:
Adding an acceptance test to the monorepo here https://github.com/ethereum-optimism/optimism/pull/18724.

joshklop commented on internal/ethapi/simulate.go:
```diff
@@ -177,7 +178,9 @@ type simulator struct {
-func (sim *simulator) execute(ctx context.Context, blocks []simBlock) ([]*simBlockResult, error) {
+// OPStack-diff: execute accepts an l1 attributes transaction which (if non-nil) will be injected into each block
+// at position 0.
+func (sim *simulator) execute(ctx context.Context, blocks []simBlock, l1AttributesTx *types.Transaction) ([]*simBlockResult, error) {
```
Looks like the `l1AttributesTx` is passed along with the `simulator` object everywhere. Could it be a field on `simulator` instead to avoid the cascading diffs?

geoknee replied:
yes I like that [130a60f](https://github.com/ethereum-optimism/op-geth/pull/732/commits/130a60f134bb4643750b68534ff8748f50bb7c7b)

joshklop commented on internal/ethapi/simulate.go:
```diff
@@ -359,12 +364,36 @@ func (sim *simulator) processBlock(ctx context.Context, block *simBlock, header,
+	// expects the first transaction to be a deposit transaction containing L1 attributes data.
+	isOptimism := sim.chainConfig.IsOptimism()
+	finalTxes := txes
+	if isOptimism {
```
Can the receipt be injected here as well to avoid rebuilding the block without the `l1AttributesTx` below?

geoknee replied:
I think we need to rebuild the block without the l1AttributesTx in order to maintain the expected behavior for users (they pass one transaction to simulate in a block, they get back a block with one transaction). I imagine this is was most tooling is going to expect that.

geoknee replied:
@kustrun can you confirm?

kustrun replied:
Yes, rebuilding the block without `l1AttributesTx` seems like the right move. It maintains backward compatibility, avoids breaking existing tooling, and prevents serialization issues for users of `go-ethereum`, which doesn’t recognize L2 system transactions unless `op-geth` is used.

joshklop replied:
Ah missed that, thanks.

joshklop commented on internal/ethapi/api.go:
```diff
@@ -921,7 +921,18 @@ func (api *BlockChainAPI) SimulateV1(ctx context.Context, opts simOpts, blockNrO
+			return nil, err
+		}
+		if len(latestBlock.Transactions()) == 0 {
+			return nil, errors.New("optimism: no L1 attributes transaction found in latest block")
```
Thoughts on special-casing the genesis block here? This would mean some changes lower in the call stack will probably change into `IsOptimism == true && sim.l1AttributesTx != nil`

geoknee replied:
Are you suggesting we remove the early return here and allow the preexisting error to bubble up for the genesis block case? I think that is a bit cleaner, yes c421c4f86478be87a017048036175fc11fcb7991.

joshklop replied:
I actually meant implementing `eth_simulateV1` to work for the genesis block as well, but no worries.

geoknee committed:
Embed L1 attributes tx in simulator

geoknee committed:
Handle Optimism genesis without L1 tx

geoknee committed:
Rename finalTxes to prependedTxes

geoknee committed:
Remove nil argument from sim.execute calls

kustrun commented on internal/ethapi/api.go:
```diff
@@ -921,6 +921,18 @@ func (api *BlockChainAPI) SimulateV1(ctx context.Context, opts simOpts, blockNrO
 		validate:       opts.Validation,
 		fullTx:         opts.ReturnFullTransactions,
 	}
+	if api.b.ChainConfig().IsOptimism() {
+		latestBlock, err := api.b.BlockByNumber(ctx, rpc.LatestBlockNumber)
+		if err != nil {
+			return nil, err
+		}
+		if len(latestBlock.Transactions()) > 0 {
+			// For the special case of the genesis block, there are no transactions
+			// so we won't set the l1AttributesTx and the simulation will
+			// likely fail.
+			sim.l1AttributesTx = latestBlock.Transactions()[0]
+		}
+	}
```
Would it make sense to instead fetch the 0th transaction of the **state block**? That way, the simulation would be unambiguous and consistently work even for historical blocks (and report correct values for historical blobs like you mentioned).

```suggestion
	if api.b.ChainConfig().IsOptimism() {
		block, err := api.b.BlockByNumberOrHash(ctx, *blockNrOrHash)
		if err != nil {
			return nil, err
		}
		if len(block.Transactions()) > 0 {
			// For the special case of the genesis block, there are no transactions
			// so we won't set the l1AttributesTx and the simulation will
			// likely fail.
			sim.l1AttributesTx = block.Transactions()[0]
		}
	}
```

geoknee replied:
Yes, thank you, this makes sense. Will apply your patch.

geoknee committed:
Update internal/ethapi/api.go

Co-authored-by: kustrun <9192608+kustrun@users.noreply.github.com>

kustrun reviewed (commented):
Tested the fix on the live node, and it’s resolving the bug. I really like the clean, minimal solution we arrived at!
